spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/ExpireSnapshotsSparkAction.java

```diff
@@ -184,11 +187,21 @@ public Dataset<FileInfo> expireFiles() {
 
       expireSnapshots.cleanExpiredFiles(false).commit();
 
-      // fetch metadata after expiration
-      Dataset<FileInfo> validFileDS = validFileDS(ops.refresh());
+      // determine expired files
+      TableMetadata updatedTable = ops.refresh();
+      Set<Long> retainedSnapshots =
```

Comment:
What about adding `findExpiredSnapshotIds` that would accept `originalMetadata`, `updatedMetadata`?

```
// fetch valid files after expiration
TableMetadata updatedMetadata = ops.refresh();
Dataset<FileInfo> validFileDS = fileDS(updatedMetadata);

// fetch files referenced by expired snapshots
Set<Long> expiredSnapshotIds = findExpiredSnapshotIds(originalMetadata, updatedMetadata);
Dataset<FileInfo> deleteCandidateFileDS = fileDS(originalMetadata, expiredSnapshotIds);

// determine expired files
this.expiredFileDS = deleteCandidateFileDS.except(validFileDS);
```

Reply:
Done